### ISSUE TITLE
Download percentage stuck at 98% or 99%

### ISSUE THREAD

**Issue by borrelnoten** [alanmcgovern/monotorrent#368]
Using latest unstable gnu package.
Download stuck at 98 or 99. Download rate at 600bytes at that point. Tested with 4 magnet urls.

**Comment by borrelnoten** [comment]
looked more into this. Seems the last few bytes just do not get downloaded.


I check if a file is compleeted like this:

f.BytesDownloaded() >= f.Length

I read the progress of the file download like this:

(int)torrentStreamProvider.Files[cur_torFileIndexNum].BitField.PercentComplete

the example is stuck at 99%


Example magnet:

magnet:?xt=urn:btih:0f57761454b47ee53e263da475d4b4cba5d549a6&dn=The.Mandalorian.S02E08.1080p.WEB.H264-DRYB%5Brartv%5D&tr=http%3A%2F%2Ftracker.trackerfix.com%3A80%2Fannounce&tr=udp%3A%2F%2F9.rarbg.me%3A2990&tr=udp%3A%2F%2F9.rarbg.to%3A2770

**Comment by alanmcgovern** [comment]
@borrelnoten Can you tell me the exact version you're using?  Is it `v2.0.0-alpha.unstable.rev0031` ?

**Comment by borrelnoten** [comment]
17 ... last one instable zaterdag, 19 december 2020, 02:18p.m. +01:00 van Alan McGovern  notifications@github.com :

**Comment by alanmcgovern** [comment]
Also - when the torrent gets stuck could you tell me if `TorrentManager.IsInEndGame` returns `false` ? I wonder if there was an unexpected sequence of events which meant that some pieces weren't being downloaded and then we didn't reach the threshold to enter endgame mode where the remaining pieces are downloaded from multiple peers concurrently.

I've been tweaking the exact semantics of when peers are allowed download blocks from the same piece, and i think the new semantics are a bit too restrictive. I've been adding some additional cases to this which should help! :) 

I have a new idea for how to make this even easier to manage in future. The current logic is a wee bit complicated but With This One Small Change i could simplify it!

The tl;dr is that when a peer selects a brand new piece to download, we populate the 'RequestedOff' field with that peer. At a later point when we actually request a block of data we'll set the `Requested` flag to true.

If something happens to a peer (disconnected, or it chokes us, or it stops sending messages, or whatever) I can null out the `RequestedOff` field and then all other peers will be allowed request the piece. This will (hopefully) be a slightly simpler way of managing the current logic, but shouldn't have any impact on actual performance in comparison to what's in master right now... unless i discover a bug :)

**Comment by alanmcgovern** [comment]
@borrelnoten Ah, excellent! The newer release, rev 31, has this fix:

> Improved handling for pieces which are partially downloaded by one peer who disconnects before the piece has completed.

I think this could be the fix for the problem you described!

**Comment by borrelnoten** [comment]
I'll test with .31 version today.

Side question, when I have a magnet URL, and I want to add trackers, I currently add the trackers url to the magnet url like this. Is this the way to do it?

 // add trackers to the magnet uri
                        string rawTrackerList = torsets.TrackersCSV;
                        string trackerList = "";

                        foreach (string tr in Regex.Replace(rawTrackerList.Replace(System.Environment.NewLine, ""), @"\s+", "", RegexOptions.Compiled).Split(','))
                        {
                            trackerList += "&tr=" + HttpUtility.UrlEncode(tr);
                        }

                        var mlink = MagnetLink.FromUri(new Uri(location));
                        var manager = new TorrentManager(mlink, GetDownloaderData().FullDownloaderPath, torrentDefaults, Path.Combine(GetDownloaderData().FullDownloaderPath, "tmp.torrent"));

                        torrent = manager.Torrent;

                        manager.Dispose();

**Comment by alanmcgovern** [comment]
I just issued a new release, rev 37, which has further tweaks to how piece picking is performed. It shouldn't be too much different to rev31, but testing with rev37 would be preferred!

`public TorrentManager (InfoHash infoHash, string savePath, TorrentSettings settings, string torrentSave, IList<IList<string>> announces)`

If your magnet link doesn't have trackers, you could use this constructor overload to add a list of trackers manually.

You could also call `TorrentManager.TrackerManager.AddTrackerAsync` to add another tracker to a torrent manager. If you do this before the torrent starts it will be used immediately after you call `StartAsync`. If you add it while the torrent is running you have to wait for the regular announce interval to be reached before the tracker will be queried (iirc).

**Comment by borrelnoten** [comment]
I'll test 37 version today.

On the tracker question: would the torrent have the additional tracker Uri's if I would use the trackermanager like this:

1: create the torrentmanager with the magnet url
2: add additional tracker Uri's async
3: get the torrent object -> this I will use later.
4: dispose the manager

                        var mlink = MagnetLink.FromUri(new Uri(location));
                        var manager = new TorrentManager(mlink, GetDownloaderData().FullDownloaderPath, torrentDefaults, Path.Combine(GetDownloaderData().FullDownloaderPath, "tmp.torrent"));

                        foreach (string tr in Regex.Replace(rawTrackerList.Replace(System.Environment.NewLine, ""), @"\s+", "", RegexOptions.Compiled).Split(','))
                        {
                            manager.TrackerManager.AddTrackerAsync(new Uri(tr));
                        }

                        torrent = manager.Torrent;

**Comment by alanmcgovern** [comment]
No - if you wanted to do that you'd have to use TorrentBuilder (iirc) to modify the underlying data. You could save the modified data to disk and then you'll have the additional trackers if you load that into the engine instead of using the magnet link.

**Comment by borrelnoten** [comment]
ah oke, will do that!

**Comment by borrelnoten** [comment]
Tested with 20 different torrents. none of them are now stuck at the end anymore and properly finish downloading.

**Comment by borrelnoten** [comment]
will keep monitoring

**Comment by alanmcgovern** [comment]
Great! Thanks!
